src/client/Microsoft.Identity.Client/ManagedIdentity/AppServiceManagedIdentitySource.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Internal;
+using Microsoft.Identity.Client.Utils;
+
+namespace Microsoft.Identity.Client.ManagedIdentity
+{
+    /// <summary>
+    /// Original source of code: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/src/AppServiceManagedIdentitySource.cs
+    /// </summary>
+    internal class AppServiceManagedIdentitySource : ManagedIdentitySource
+    {
+        // MSI Constants. Docs for MSI are available here https://docs.microsoft.com/azure/app-service/overview-managed-identity
+        private const string AppServiceMsiApiVersion = "2019-08-01";
+        private const string SecretHeaderName = "X-IDENTITY-HEADER";
+
+        private const string MsiEndpointInvalidUriError = "[Managed Identity] The environment variable IDENTITY_ENDPOINT contains an invalid Uri {0} in app service managed identity source.";
+
+        private readonly Uri _endpoint;
+        private readonly string _secret;
+        private readonly string _clientId;
+        private readonly string _resourceId;
+
+        public static ManagedIdentitySource TryCreate(RequestContext requestContext)
+        {
+            var msiSecret = EnvironmentVariables.IdentityHeader;
+
+            return TryValidateEnvVars(EnvironmentVariables.IdentityEndpoint, msiSecret, requestContext.Logger, out Uri endpointUri)
+                ? new AppServiceManagedIdentitySource(requestContext, endpointUri, msiSecret)
+                : null;
+        }
+
+        private AppServiceManagedIdentitySource(RequestContext requestContext, Uri endpoint, string secret) : base(requestContext)
+        {
+            _endpoint = endpoint;
+            _secret = secret;
+            _clientId = requestContext.ServiceBundle.Config.MIUserAssignedClientId;
+            _resourceId = requestContext.ServiceBundle.Config.MIUserAssignedResourceId;
+        }
+
+        private static bool TryValidateEnvVars(string msiEndpoint, string secret, ILoggerAdapter logger, out Uri endpointUri)
+        {
+            endpointUri = null;
+
+            // if BOTH the env vars endpoint and secret values are null, this MSI provider is unavailable.
+            if (string.IsNullOrEmpty(msiEndpoint) || string.IsNullOrEmpty(secret))
+            {
+                logger.Info("[Managed Identity] App service managed identity is unavailable.");
```

Comment:
nit: I think this one should be Verbose and the success should be Info because otherwise, we may end up printing out too many "not found" messages.